### ISSUE TITLE
Compare to other enum proposals?

### ISSUE THREAD

**Issue by ljharb** [Jack-Works/proposal-enum#22]
How does this compare to @rwaldron's https://github.com/rwaldron/proposal-enum-definitions and @rbuckton's https://github.com/rbuckton/proposal-enum ?

**Comment by Jack-Works** [comment]
I have communicated with @rbuckton. I'll revisit the materials see if I can cover the support of shared structs (which @rbuckton's version indented to support).
After revisiting the materials I'll update the documentation.

**Comment by rwaldron** [comment]
@ljharb thanks for the call out. My efforts have largely merged with @rbuckton's, with only a few minor nits that I think could be worked out.

**Comment by ljharb** [comment]
See https://github.com/rbuckton/proposal-enum/issues/1 for some discussion that may be related to why neither prior proposal has been brought to plenary yet.

**Comment by ljharb** [comment]
Id love to see some resolution on things like https://github.com/rbuckton/proposal-enum/issues/1 ; i haven’t gotten any sense of that merging, but perhaps it’s happened privately?

**Comment by rwaldron** [comment]
@ljharb it was mostly resolved off-line and I believe resulted in @rbuckton's `Foo of Bar` additions (specifically the `enumMap` mechanism, IIRC in https://github.com/rbuckton/proposal-enum/commit/f3a62f266e8807f29b1e916428b63ea62c39d7d2), which was _close_ but not quite the novel thing I described with the `Foo via Bar` design. That difference is what I thought we could work out over time. Unfortunately, the whole effort stalled when at least one committee member vowed to block the proposal of any enum design that had default integer values 🤷

**Comment by rwaldron** [comment]
@ljharb at this point, I could probably be convinced to go along with "no default incrementing integer values", but only if the `enumMap` mechanism is included. This proposal is already trying to special case the creation of the values (by highjacking the syntactic space that Ron and I were working in), instead of allowing it to be a user-provided mechanism, which I can't support. 

I don't want this: 

```js
enum Days for bigint {
    Monday = 0n,
    Friday = 4n,
}
```

I want this: 

```js
enum Days of BigInt {
    Monday,
    Friday = 4n,
}
```

Which gives me: `Days.Monday === 0n` and `Days.Friday === 4n`. But given no specific assignments: 

```js
enum Days of BigInt {
    Monday,
    Tuesday, 
    Wednesday,
    Thursday,
    Friday,
    Saturday, 
    Sunday
}
```

Would give me: 

```js
Days.Monday === 0n;
Days.Tuesday  === 1n;
Days.Wednesday === 2n;
Days.Thursday === 3n;
Days.Friday === 4n;
Days.Saturday  === 5n;
Days.Sunday === 6n;
```


And if I'm going to make good on my commitment to compromise, then I would expect the following to fail: 

```js
enum Days {
    Monday
}
```

```
SyntaxError: missing assignment after Monday
```
(or something like that)


This capability is a hard requirement.

**Comment by rbuckton** [comment]
I'd like to point out the `@@toEnum` approach in my proposal is heavily influenced by Python enums, but folds in Python's `auto()` to be built in to the syntax.

I've also been working out how to extend the syntax to support ADTs based on a conversation with Axel Rauchmeyer on Twitter a few months back. An early sketch of the idea is here: https://gist.github.com/rbuckton/4a5108fab40ac90551bf82d9884711b5

**Comment by rwaldron** [comment]
@rbuckton whoa—I like that!

**Comment by Jack-Works** [comment]
> Unfortunately, the whole effort stalled when at least one committee member vowed to block the proposal of any enum design that had default integer values 🤷

Yeah, I also think auto-increment is a footgun although it is very useful. See explain in https://github.com/Jack-Works/proposal-enum/discussions/13

**Comment by rbuckton** [comment]
I wrote up another example, this time using normal classes/functions (though I think a solution that works with records/tuples/shared structs will be important in the long run):

https://gist.github.com/rbuckton/192c2922650e05a1ca9cd7c01be7fc6c

**Comment by rbuckton** [comment]
> And if I'm going to make good on my commitment to compromise, then I would expect the following to fail:
> 
> ```js
> enum Days {
>     Monday
> }
> ```

What if the default for `enum Days` was `of ADT`? Its not TS compatible (and thus would require a flag), but it would simplify the more powerful ADT-style enum case. The downside would be that non-complex enum members would probably still have a `typeof` of `"function"`.

**Comment by Jack-Works** [comment]
I have updated the README. I didn't include the comparison to the previous work on this topic because I decided not to start designing concrete semantics/syntax before we figure out the design constraints.

**Comment by joshgrift** [comment]
@rwaldron @Jack-Works and @rbuckton 
I'd love to help push this forward, but there are a few different proposals floating around. It sounds like @rbuckton and @rwaldron are close to an agreement on the specs? What differences still remain?

**Comment by rwaldron** [comment]
@joshgrift yes, I would describe my position as close agreement with @rbuckton's latest proposal, with a few minor things to work out.
